flow/workflows/maintenance_flow.go

```diff
@@ -0,0 +1,275 @@
+	temporalClient client.Client,
+	input *protos.EndMaintenanceFlowInput,
+	taskQueueId shared.TaskQueueID,
+) (client.WorkflowRun, error) {
+	startWorkflowOptions := client.StartWorkflowOptions{
+		// This is to ensure that maintenance workflows are deduped
+		WorkflowIDReusePolicy:    tEnums.WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE,
+		WorkflowIDConflictPolicy: tEnums.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING,
+		TaskQueue:                peerdbenv.PeerFlowTaskQueueName(taskQueueId),
+	}
+	startWorkflowOptions.ID = "end-maintenance"
+	if deploymentUid := peerdbenv.PeerDBDeploymentUID(); deploymentUid != "" {
+		startWorkflowOptions.ID += "-" + deploymentUid
+	}
+
+	workflowRun, err := temporalClient.ExecuteWorkflow(ctx, startWorkflowOptions, EndMaintenanceWorkflow, &protos.EndMaintenanceFlowInput{})
+	if err != nil {
+		return nil, err
+	}
+	return workflowRun, nil
+}
+
+func StartMaintenanceWorkflow(ctx workflow.Context, input *protos.StartMaintenanceFlowInput) (*protos.StartMaintenanceFlowOutput, error) {
+	logger := workflow.GetLogger(ctx)
+	logger.Info("Starting StartMaintenance workflow", "input", input)
+	defer runBackgroundAlerter(ctx)()
+
+	maintenanceFlowOutput, err := startMaintenance(ctx, logger)
+	if err != nil {
+		slog.Error("Error in StartMaintenance workflow", "error", err)
+		return nil, err
+	}
+	return maintenanceFlowOutput, nil
+}
+
+func startMaintenance(ctx workflow.Context, logger log.Logger) (*protos.StartMaintenanceFlowOutput, error) {
+	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+		StartToCloseTimeout: 24 * time.Hour,
+	})
+
+	snapshotWaitCtx := workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+		StartToCloseTimeout: 24 * time.Hour,
+		HeartbeatTimeout:    1 * time.Minute,
+	})
+	waitSnapshotsFuture := workflow.ExecuteActivity(snapshotWaitCtx,
+		maintenance.WaitForRunningSnapshots,
+	)
+	err := waitSnapshotsFuture.Get(snapshotWaitCtx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	enableMaintenanceFuture := workflow.ExecuteActivity(ctx, maintenance.EnableMaintenanceMode)
+	err = enableMaintenanceFuture.Get(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("Waiting for all snapshot mirrors to finish snapshotting")
+	waitSnapshotsPostEnableFuture := workflow.ExecuteActivity(snapshotWaitCtx,
+		maintenance.WaitForRunningSnapshots,
+	)
+	err = waitSnapshotsPostEnableFuture.Get(snapshotWaitCtx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	mirrorsList, err := getAllMirrors(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	runningMirrors, err := pauseAndGetRunningMirrors(ctx, mirrorsList, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	future := workflow.ExecuteActivity(ctx, maintenance.BackupAllPreviouslyRunningFlows, runningMirrors)
+	err = future.Get(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	logger.Info("StartMaintenance workflow completed")
+	return &protos.StartMaintenanceFlowOutput{
+		Version: peerdbenv.PeerDBVersionShaShort(),
```

Comment:
This needs to be wrapped in a local activity. See `workflows/local_activities.go` for how other workflows access environment variables